horovod/_keras/__init__.py

```diff
@@ -60,22 +40,42 @@ def get_gradients(self, loss, params):
             In DistributedOptimizer, get_gradients() is overriden to also
             allreduce the gradients before returning them.
             """
+            self._fail_if_executing_eagerly()
             gradients = super(self.__class__, self).get_gradients(loss, params)
             if hvd.size() > 1:
-                return self._allreduce_grads(gradients)
+                averaged_gradients = []
+                with tf.name_scope(self._name + "_Allreduce"):
+                    for grad in gradients:
+                        if grad is not None:
+                            if self._sparse_as_dense and \
+                                    isinstance(grad, tf.IndexedSlices):
+                                grad = tf.convert_to_tensor(grad)
+                            avg_grad = hvd.allreduce(grad,
+                                                     device_dense=self._device_dense,
+                                                     device_sparse=self._device_sparse,
+                                                     compression=self._compression)
+                            averaged_gradients.append(avg_grad)
+                        else:
+                            averaged_gradients.append(None)
+                    return averaged_gradients
             else:
                 return gradients
 
+        def apply_gradients(self, *args, **kwargs):
+            self._fail_if_executing_eagerly()
```

Comment:
Is it possible to just have the eager failure occur during `__init__` instead of in both `apply_gradients` and `compute_gradients`?

Reply:
My understanding is that at the time the optimizer is created it's fine to be running eagerly, only during fit you need to be in graph mode.

Reply:
Yeah, that's exactly right.

Reply:
I figured that might be the case but wanted to check. Thanks!